Applications/Spire/Source/Ui/OverlayPanel.cpp

```diff
@@ -45,6 +45,55 @@ namespace {
   }
 }
 
+class OverlayPanel::ParentMoveObserver : public QObject {
+  public:
+    using MoveSignal = Signal<void()>;
+
+    ParentMoveObserver(QWidget& widget) {
+      install_parent_event_filter(widget);
+    }
+
+    ~ParentMoveObserver() {
+      for(auto parent : m_parents) {
+        parent->removeEventFilter(this);
+      }
+    }
+
+    connection connect_move_signal(const MoveSignal::slot_type& slot) const {
+      return m_move_signal.connect(slot);
+    }
+
+    bool eventFilter(QObject* watched, QEvent* event) override {
+      if(event->type() == QEvent::Move) {
+        m_move_signal();
+      } else if(event->type() == QEvent::ParentChange) {
+        auto i = std::find(m_parents.begin(), m_parents.end(), watched);
+        auto j = std::next(i, 1);
+        while(j != m_parents.end()) {
+          (*j)->removeEventFilter(this);
+          j = m_parents.erase(j);
+        }
+        if(i != m_parents.end()) {
+          install_parent_event_filter(*(*i));
+        }
+      }
+      return QObject::eventFilter(watched, event);
+    }
+
+  private:
+    mutable MoveSignal m_move_signal;
+    std::list<QWidget*> m_parents;
```

Comment:
Could you use a std::unordered_set for this instead? It looks like it's iterated over, searched, and written to, so the search could become constant while still supporting the other operations.

Reply:
The parents are stored in the list in order from the most direct parent to the top-level parent. When the parent of a widget stored in the list (m_parents) changes, all parents behind the widget in the list need to be removed and new parents of the widget need to be added in order. 

For example, suppose m_parents has a list of [`parent1`, `parent2`, `parent3`, `parent4`], `parent2` is the parent of `parent1`, and `parent3` is the parent of `parent2`, …  and `parent4` is the top-level parent. When the parent of `parent2` changes, `parent3` and `parent4` need to be removed from the list, and new parents are added after `parent2`, like [`parent1`, `parent2`, `new-parent3`, `new-parent4`, `new-parent5`].

`std::unordered_set` can’t store values in order, so it is not a suitable container. Another alternative is `std::vector`, but the list might need to erase and add items multiple times, so I think `std::list` is better than `std::vector`.